src/Nethermind/Nethermind.Optimism/OptimismPlugin.cs

```diff
@@ -0,0 +1,326 @@
+        ArgumentNullException.ThrowIfNull(_api.DbProvider);
+        ArgumentNullException.ThrowIfNull(_api.BlockTree);
+        ArgumentNullException.ThrowIfNull(_api.ReadOnlyTrieStore);
+        ArgumentNullException.ThrowIfNull(_api.BlockValidator);
+        ArgumentNullException.ThrowIfNull(_api.RewardCalculatorSource);
+        ArgumentNullException.ThrowIfNull(_api.ReceiptStorage);
+        ArgumentNullException.ThrowIfNull(_api.TxPool);
+        ArgumentNullException.ThrowIfNull(_api.TransactionComparerProvider);
+
+        _api.BlockProducerEnvFactory = new OptimismBlockProducerEnvFactory(
+            _api.ChainSpec,
+            _api.DbProvider,
+            _api.BlockTree,
+            _api.ReadOnlyTrieStore,
+            _api.SpecProvider,
+            _api.BlockValidator,
+            _api.RewardCalculatorSource,
+            _api.ReceiptStorage,
+            _api.BlockPreprocessor,
+            _api.TxPool,
+            _api.TransactionComparerProvider,
+            _blocksConfig,
+            _api.LogManager);
+
+        _api.GasLimitCalculator = new OptimismGasLimitCalculator();
+
+        _blockProductionTrigger = new BuildBlocksWhenRequested();
+
+        BlockProducerEnv producerEnv = _api.BlockProducerEnvFactory.Create();
+
+        _api.BlockProducer = _blockProducer = new OptimismPostMergeBlockProducer(
+            new OptimismPayloadTxSource(),
+            producerEnv.TxSource,
+            producerEnv.ChainProcessor,
+            producerEnv.BlockTree,
+            _blockProductionTrigger,
+            producerEnv.ReadOnlyStateProvider,
+            _api.GasLimitCalculator,
+            NullSealEngine.Instance,
+            new ManualTimestamper(),
+            _api.SpecProvider,
+            _api.LogManager,
+            _api.Config<IBlocksConfig>());
+
+        return Task.FromResult(_api.BlockProducer);
+    }
+
+    #endregion
+
+    public INethermindApi CreateApi(IConfigProvider configProvider, IJsonSerializer jsonSerializer,
+        ILogManager logManager, ChainSpec chainSpec) =>
+        new OptimismNethermindApi(configProvider, jsonSerializer, logManager, chainSpec);
+
+    public Task Init(INethermindApi api)
+    {
+        if (!ShouldRunSteps(api))
+            return Task.CompletedTask;
+
+        _api = (OptimismNethermindApi)api;
+        _mergeConfig = _api.Config<IMergeConfig>();
+        _syncConfig = _api.Config<ISyncConfig>();
+        _blocksConfig = _api.Config<IBlocksConfig>();
+        _logger = _api.LogManager.GetClassLogger();
+
+        ArgumentNullException.ThrowIfNull(_api.BlockTree);
+
+        _api.PoSSwitcher = AlwaysPoS.Instance;
+
+        _blockCacheService = new BlockCacheService();
+        _api.InvalidChainTracker = _invalidChainTracker = new InvalidChainTracker(
+            _api.PoSSwitcher,
+            _api.BlockTree,
+            _blockCacheService,
+            _api.LogManager);
+        _api.DisposeStack.Push(_invalidChainTracker);
```

Comment:
It was very tricky to do in InitBlockchain because of dependency loop. I decided to move SetupBlockchainProcessorInterceptor call to OptimismPlugin like it's done in MergePlugin. I can try to move it, but it's gonna be pretty ugly probably

Reply:
Can you clarify on dependency loop?